### ISSUE TITLE
NewGVKParser fails when an aggregated api server encouters types of same name but different version

### ISSUE THREAD

**Issue by kevindelgado** [kubernetes/kubernetes#103597]
#### What happened:

Attempting to create a new GVKParser fails with
```
duplicate entry for /v1, Kind=APIResourceList
```

Upon looking at the output of `/openapi/v2` I see that the OpenAPIDoc has an entry for both `"io.k8s.apimachinery.pkg.apis.meta.v1.APIResourceList"` and `"io.k8s.apimachinery.pkg.apis.meta.v1.APIResourceList_v2"`.

The only difference between them is the link to the documentation in the description of "apiVersion" property. The only reference to `_v2` is from `"/apis/metrics.k8s.io/v1beta1/"` meaning the metrics server is on a different version of kubernetes than the kube-apiserver. 

#### Anything else we need to know?:

When the kube-aggregator aggregates the openAPI from it’s subcomponents, it sometimes encounters two types that have the same name and different definitions. It resolves this by appending _v2, etc to the second definition, which resolves the issue for the openAPI doc, but then the GVKParser does nothing to disambiguate these gvks and will fail if NewGVKParser is ever called.

#### How to reproduce it (as minimally and precisely as possible):
Given a valid kubeconfig:
 ```
 dc := discovery.NewDiscoveryClientForConfigOrDie(config)
 doc, err := dc.OpenAPISchema()
 models, err := proto.NewOpenAPIData(doc)
 gvkParser, err := fieldmanager.NewGVKParser(models, false)
 if err != nil {
     panic(err) // this will panic with `duplicate entry for /v1, Kind=APIResourceList`
 }
 ```

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.6-gke.1000
- Cloud provider or hardware configuration: GKE

**Comment by kevindelgado** [comment]
cc @apelisse

**Comment by lavalamp** [comment]
cc @deads2k

**Comment by leoluz** [comment]
We are facing this issue using the `fieldmanager.NewGVKParser()` in ArgoCD. 
Is there a workaround for this issue?

**Comment by apelisse** [comment]
Hard to tell without knowing more about the actual set-up. Also, we should have a work-around internally to avoid that issue.

**Comment by apelisse** [comment]
Do you have an aggregated apiserver that isn't at the same version as the apiserver?

**Comment by leoluz** [comment]
> Do you have an aggregated apiserver that isn't at the same version as the apiserver?

As I am invoking `fieldmanager.NewGVKParser()` inside ArgoCD code base I guess that there is a possibility that users will have that condition. Currently I am seeing this error inside our e2e tests for example. I guess I need to somehow handle this error gracefully.

**Comment by liggitt** [comment]
note that the _v2 objects show up even when there's no aggregated server... this is a bug internal to kube-apiserver tracked in https://github.com/kubernetes/kubernetes/issues/109275

**Comment by apelisse** [comment]
Yeah, and even if an aggregated apiservers has that condition, we probably shouldn't break the world.

**Comment by crenshaw-dev** [comment]
I'm seeing what appears to be a regression in 1.30. This is what I get with Rancher running 1.30.0:

```
kubectl get --raw /openapi/v2 | jq -r '.definitions | keys | .[]' | sort | grep io.k8s.apimachinery.pkg.apis.meta.v1
io.k8s.apimachinery.pkg.apis.meta.v1.APIGroup
io.k8s.apimachinery.pkg.apis.meta.v1.APIGroupList
io.k8s.apimachinery.pkg.apis.meta.v1.APIResource
io.k8s.apimachinery.pkg.apis.meta.v1.APIResourceList
io.k8s.apimachinery.pkg.apis.meta.v1.APIResourceList_v2
io.k8s.apimachinery.pkg.apis.meta.v1.APIResource_v2
io.k8s.apimachinery.pkg.apis.meta.v1.APIVersions
io.k8s.apimachinery.pkg.apis.meta.v1.Condition
io.k8s.apimachinery.pkg.apis.meta.v1.DeleteOptions
io.k8s.apimachinery.pkg.apis.meta.v1.Duration
io.k8s.apimachinery.pkg.apis.meta.v1.FieldsV1
io.k8s.apimachinery.pkg.apis.meta.v1.GroupVersionForDiscovery
io.k8s.apimachinery.pkg.apis.meta.v1.LabelSelector
io.k8s.apimachinery.pkg.apis.meta.v1.LabelSelectorRequirement
io.k8s.apimachinery.pkg.apis.meta.v1.ListMeta
io.k8s.apimachinery.pkg.apis.meta.v1.ManagedFieldsEntry
io.k8s.apimachinery.pkg.apis.meta.v1.MicroTime
io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta
io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta_v2
io.k8s.apimachinery.pkg.apis.meta.v1.OwnerReference
io.k8s.apimachinery.pkg.apis.meta.v1.Patch
io.k8s.apimachinery.pkg.apis.meta.v1.Preconditions
io.k8s.apimachinery.pkg.apis.meta.v1.ServerAddressByClientCIDR
io.k8s.apimachinery.pkg.apis.meta.v1.Status
io.k8s.apimachinery.pkg.apis.meta.v1.StatusCause
io.k8s.apimachinery.pkg.apis.meta.v1.StatusDetails
io.k8s.apimachinery.pkg.apis.meta.v1.Time
io.k8s.apimachinery.pkg.apis.meta.v1.WatchEvent
```

**Comment by crenshaw-dev** [comment]
The k8s definition of `APIResourceList` is out of sync with the metrics-server definition of `APIResourceList` as of metrics-server v 0.7.1 (released in March). This [fix](https://github.com/kubernetes-sigs/metrics-server/commit/19f9e22179dd4bc3c846f3655e1f38de03c0f27d) from May would solve the duplicate definition problem and allow the NewGVKParser to be constructed. 

One solution is to wait for a new metrics-server release. What I'll probably do in Argo CD is add a definition pre-filter to detect duplicate definitions and drop one, logging a warning. Should be good enough for most situations.
